src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs

```diff
@@ -17175,5 +17184,347 @@ static void Main(List<int> list)
+
+                    static void Main()
+                    {
+                        F([]);
+                    }
+                }
+                """;
+
+            CreateCompilation(source).VerifyEmitDiagnostics(
+                // (18,11): error CS1503: Argument 1: cannot convert from 'collection expressions' to 'NoConstructorCollection'
+                //         F([]);
+                Diagnostic(ErrorCode.ERR_BadArgType, "[]").WithArguments("1", "collection expressions", "NoConstructorCollection").WithLocation(18, 11)
+                );
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/69521")]
+        public void MissingCtor_OverloadResolution()
+        {
+            string source = """
+                using System.Collections;
+                using System.Collections.Generic;
+
+                abstract class AbstractCollection : IEnumerable<int>
+                {
+                    IEnumerator IEnumerable.GetEnumerator() => null;
+                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => null;
+                    public void Add(int i) { }
+                }
+
+                class NoConstructorCollection : IEnumerable<int>
+                {
+                    private NoConstructorCollection() { }
+                    IEnumerator IEnumerable.GetEnumerator() => null;
+                    IEnumerator<int> IEnumerable<int>.GetEnumerator() => null;
+                    public void Add(int i) { }
+                }
+
+                class Program
+                {
+                    static void F(AbstractCollection c) { }
+                    static void F(NoConstructorCollection c) { }
+                    static void F(List<int> c) { }
+
+                    static void Main()
+                    {
+                        F([]);
+                    }
+                }
+                """;
+
+            var comp = CreateCompilation(source).VerifyEmitDiagnostics();
+
+            var tree = comp.SyntaxTrees.Single();
+            var model = comp.GetSemanticModel(tree);
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            Assert.Equal("F([])", invocation.ToString());
+
+            Assert.Equal("void Program.F(System.Collections.Generic.List<System.Int32> c)",
+                model.GetSymbolInfo(invocation).Symbol.ToTestDisplayString());
+        }
+
+        [CombinatorialData]
+        [Theory]
+        public void CollectionBuilder_MissingCtor(bool useCompilationReference)
+        {
+            string sourceA = """
+                using System;
+                using System.Collections;
+                using System.Collections.Generic;
+                using System.Runtime.CompilerServices;
+                [CollectionBuilder(typeof(MyCollectionBuilder), nameof(MyCollectionBuilder.Create))]
+                public struct MyCollection<T> : IEnumerable<T>
```

Comment:
>struct [](http://example.com/codeflow?start=23&length=6)

Are we testing that the constructor requirement is ignored for `[CollectionBuilder]` types?

If so, this should probably be a `class` instead since the `struct` type has a public parameterless constructor.

Also, we could probably limit the examples in the test to just one case, such as `MyCollection<object> c = [1, 2, 3];`.